src/git-repository-async.js

```diff
@@ -0,0 +1,871 @@
+      this.subscriptions.add(this.project.onDidAddBuffer(buffer => this.subscribeToBuffer(buffer)))
+    }
+  }
+
+  // Public: Destroy this {GitRepositoryAsync} object.
+  //
+  // This destroys any tasks and subscriptions and releases the underlying
+  // libgit2 repository handle. This method is idempotent.
+  destroy () {
+    if (this.emitter) {
+      this.emitter.emit('did-destroy')
+      this.emitter.dispose()
+      this.emitter = null
+    }
+    if (this.subscriptions) {
+      this.subscriptions.dispose()
+      this.subscriptions = null
+    }
+
+    this.repoPromise = null
+  }
+
+  // Event subscription
+  // ==================
+
+  // Public: Invoke the given callback when this GitRepositoryAsync's destroy()
+  // method is invoked.
+  //
+  // * `callback` {Function}
+  //
+  // Returns a {Disposable} on which `.dispose()` can be called to unsubscribe.
+  onDidDestroy (callback) {
+    return this.emitter.on('did-destroy', callback)
+  }
+
+  // Public: Invoke the given callback when a specific file's status has
+  // changed. When a file is updated, reloaded, etc, and the status changes, this
+  // will be fired.
+  //
+  // * `callback` {Function}
+  //   * `event` {Object}
+  //     * `path` {String} the old parameters the decoration used to have
+  //     * `pathStatus` {Number} representing the status. This value can be passed to
+  //       {::isStatusModified} or {::isStatusNew} to get more information.
+  //
+  // Returns a {Disposable} on which `.dispose()` can be called to unsubscribe.
+  onDidChangeStatus (callback) {
+    return this.emitter.on('did-change-status', callback)
+  }
+
+  // Public: Invoke the given callback when a multiple files' statuses have
+  // changed. For example, on window focus, the status of all the paths in the
+  // repo is checked. If any of them have changed, this will be fired. Call
+  // {::getPathStatus(path)} to get the status for your path of choice.
+  //
+  // * `callback` {Function}
+  //
+  // Returns a {Disposable} on which `.dispose()` can be called to unsubscribe.
+  onDidChangeStatuses (callback) {
+    return this.emitter.on('did-change-statuses', callback)
+  }
+
+  // Repository details
+  // ==================
+
+  // Public: A {String} indicating the type of version control system used by
+  // this repository.
+  //
+  // Returns `"git"`.
+  getType () {
+    return 'git'
+  }
+
+  // Public: Returns a {Promise} which resolves to the {String} path of the
+  // repository.
+  getPath () {
+    return this.repoPromise.then(repo => repo.path().replace(/\/$/, ''))
+  }
+
+  // Public: Returns a {Promise} which resolves to the {String} working
+  // directory path of the repository.
+  getWorkingDirectory () {
+    return this.repoPromise.then(repo => repo.workdir())
+  }
+
+  // Public: Returns a {Promise} that resolves to true if at the root, false if
+  // in a subfolder of the repository.
+  isProjectAtRoot () {
```

Comment:
Looks like this method returns `null` instead of a `Promise` if `this.project` is falsey. Should it return a rejected Promise in that case instead?